includes/shortcodes.php

```diff
@@ -666,11 +666,27 @@ function edd_receipt_shortcode( $atts, $content = null ) {
 	}
 
 	// No key found
-	if ( ! isset( $payment_key ) )
-		return $edd_receipt_args['error'];
+	if ( ! isset( $payment_key ) ) {
+		return '<p class="edd-alert edd-alert-error">' . $edd_receipt_args['error'] . '</p>';
+	}
+
+	$payment_id    = edd_get_purchase_id_by_key( $payment_key );
+	$user_can_view = edd_can_view_receipt( $payment_key );
+
+	// Key was provided, but user is logged out. Offer them the ability to login and view the receipt
+	if ( ! $user_can_view && ! empty( $payment_key ) && ! is_user_logged_in() && ! edd_is_guest_payment( $payment_id ) ) {
```

Comment:
@cklosowski do we want the edd_user_can_view_receipt filter above this just in case?